#### 1. Hello,Data!
* Read the csv file into records.
* Prnt the first 3 rows from the csv file.
* .shape is used to verify dataset size.

In [27]:
import pandas as pd

# Load datasets with explicit file paths
records = pd.read_csv("../data/1000 Sales Records.csv")

# Quick sanity check: print dataset shapes (rows, columns)
print("Row*Column:",records.shape)

# View to first 3 rows
records.head(3)

Row*Column: (1000, 14)


Region Country   Item Type Sales Channel  \
0  Middle East and North Africa   Libya   Cosmetics       Offline   
1                 North America  Canada  Vegetables        Online   
2  Middle East and North Africa   Libya   Baby Food       Offline   

  Order Priority  Order Date   Order ID   Ship Date  Units Sold  Unit Price  \
0              M  10/18/2014  686800706  10/31/2014        8446      437.20   
1              M   11/7/2011  185941302   12/8/2011        3018      154.06   
2              C  10/31/2016  246222341   12/9/2016        1517      255.28   

   Unit Cost  Total Revenue  Total Cost  Total Profit  
0     263.33     3692591.20  2224085.18    1468506.02  
1      90.93      464953.08   274426.74     190526.34  
2     159.42      387259.76   241840.14     145419.62

In [28]:
import pandas as pd

# Lab suggest to have >=500 rows, restricts for 500 rows
records =  pd.read_csv("../data/1000 Sales Records.csv", nrows=500)

# Quick sanity check: print dataset shapes (rows, columns)
print("Row*Column:",records.shape)

Row*Column: (500, 14)


* The lab requires a dataset with at least **500 rows**, so used `nrows=500`.  

We also print the shape of the DataFrame `(rows, columns)` to confirm we are on track.

#### 2. Pick the Right Container and Justify dict vs namedtuple vs sets(1–2 sentences)

I feel the best pick for the container would be a **list of dictionaries**  as it can hold many flexible key-value pairs and also in the assignment it is mentioned to save cleaned data to JSON.
Why its best because 'namedtuple' which is constant once defined and 'set' only sotres unique values and cannot ontain all the records.

#### 3. Implement Functions and  Data structure
* Lets convert the dataframe into a *list of dictionaries*.
* Each row is converted into Python dictionary; key-value pairs.

In [29]:
def populate_data_structure(data: pd.DataFrame) -> list[dict]:
       return data.to_dict(orient="records")

# Use the function on df_raw (first 500 rows loaded earlier)
records_list = populate_data_structure(records)

print(f"Number of records: {len(records_list)}")
print("First record:", records_list[0])

Number of records: 500
First record: {'Region': 'Middle East and North Africa', 'Country': 'Libya', 'Item Type': 'Cosmetics', 'Sales Channel': 'Offline', 'Order Priority': 'M', 'Order Date': '10/18/2014', 'Order ID': 686800706, 'Ship Date': '10/31/2014', 'Units Sold': 8446, 'Unit Price': 437.2, 'Unit Cost': 263.33, 'Total Revenue': 3692591.2, 'Total Cost': 2224085.18, 'Total Profit': 1468506.02}


#### 4. Bulk Loaded
* Dataset is bulk loaded to our container (**List of Dictionary**) from DataFrame.
* Entired dataset is stored in our list.

In [30]:
# Bulk load the DataFrame into a list of dictionaries
records_list = populate_data_structure(records)

# Print confirmation
print(f"Number of records loaded: {len(records_list)}")

# Show the first 3 records as mentioned in the lab
for rec in records_list[:3]:
    print(rec)

Number of records loaded: 500
{'Region': 'Middle East and North Africa', 'Country': 'Libya', 'Item Type': 'Cosmetics', 'Sales Channel': 'Offline', 'Order Priority': 'M', 'Order Date': '10/18/2014', 'Order ID': 686800706, 'Ship Date': '10/31/2014', 'Units Sold': 8446, 'Unit Price': 437.2, 'Unit Cost': 263.33, 'Total Revenue': 3692591.2, 'Total Cost': 2224085.18, 'Total Profit': 1468506.02}
{'Region': 'North America', 'Country': 'Canada', 'Item Type': 'Vegetables', 'Sales Channel': 'Online', 'Order Priority': 'M', 'Order Date': '11/7/2011', 'Order ID': 185941302, 'Ship Date': '12/8/2011', 'Units Sold': 3018, 'Unit Price': 154.06, 'Unit Cost': 90.93, 'Total Revenue': 464953.08, 'Total Cost': 274426.74, 'Total Profit': 190526.34}
{'Region': 'Middle East and North Africa', 'Country': 'Libya', 'Item Type': 'Baby Food', 'Sales Channel': 'Offline', 'Order Priority': 'C', 'Order Date': '10/31/2016', 'Order ID': 246222341, 'Ship Date': '12/9/2016', 'Units Sold': 1517, 'Unit Price': 255.28, 'Unit

#### 5. Quick Profiling	Min/mean/max price, unique city count (set)
* Step1 : Since we dont have a city column in the csv, lets try to generate it. 
* Step2 : Post generating the column then start the objective

In [31]:
# Count unique countries
unique_country_count = records["Country"].nunique()
print(f"Number of unique countries: {unique_country_count}")

# List the unique countries
unique_countries = records["Country"].unique()
print(unique_countries)

Number of unique countries: 171
['Libya' 'Canada' 'Japan' 'Chad' 'Armenia' 'Eritrea' 'Montenegro'
 'Jamaica' 'Fiji' 'Togo' 'Greece' 'Sudan' 'Maldives' 'Estonia' 'Greenland'
 'Cape Verde' 'Senegal' 'Federated States of Micronesia' 'Bulgaria'
 'Algeria' 'Mongolia' 'Grenada' 'Mauritius ' 'Morocco' 'Honduras' 'Benin'
 'Equatorial Guinea' 'Swaziland' 'Trinidad and Tobago' 'Sweden' 'Belarus'
 'Guinea-Bissau' 'Turkey' 'Central African Republic' 'Laos' 'Israel'
 'Bhutan' 'Vanuatu' 'Burundi' 'Ukraine' 'Croatia' 'Madagascar' 'Malaysia'
 'Uzbekistan' 'Italy' 'Nepal' 'Portugal' 'Panama' 'Botswana' 'Tanzania'
 'Romania' 'Mali' 'Niger' 'Austria' 'India' 'Luxembourg' 'Iceland' 'Qatar'
 'South Sudan' 'United Kingdom' 'Tunisia ' 'United States of America'
 'Liberia' 'South Korea' 'Kenya' 'Rwanda' 'Cuba' 'Czech Republic'
 'Philippines' 'El Salvador' 'Tonga' 'Democratic Republic of the Congo'
 'Afghanistan' 'Tuvalu' 'Gabon' 'East Timor' 'Jordan' 'Cyprus' 'Malawi'
 'United Arab Emirates' 'China' 'Somalia'

In [32]:
from countryinfo import CountryInfo

def get_city(country):
    try:
        return CountryInfo(country).capital()
    except:
        return "Unknown City"
# Create shipping_city column
records["shipping_city"] = records["Country"].apply(get_city)

# Define the count of Unknown City
unknown_count = (records["shipping_city"] == "Unknown City").sum()

# Lets have the count of rows with cities and Unknown
print(f"Rows with valid city: {len(records) - unknown_count}")
print(f"Rows with 'Unknown City': {unknown_count}")

Rows with valid city: 462
Rows with 'Unknown City': 38


In [33]:
import numpy as np

# Min, mean, max of Unit Price
min_price = records["Unit Price"].min()
mean_price = records["Unit Price"].mean()
max_price = records["Unit Price"].max()

# Unique cities (using set)
unique_cities = set(records["shipping_city"])  
unique_city_count = len(unique_cities)

print(f"Min price: {min_price:.2f}")
print(f"Mean price: {mean_price:.2f}")
print(f"Max price: {max_price:.2f}")
print(f"Unique city count: {unique_city_count}")


Min price: 9.33
Mean price: 274.30
Max price: 668.27
Unique city count: 158


#### 6. Identify at least three dirty data cases

Checking for issues in:
* Check date columns are stored as object.
* Check for white space.
* Check for how many city are no available.
* Check logic between Ship date and Order Date.
* Check for duplicates.
* Check inconsistency in formating.
* Check for outliners in price.

In [34]:
# 1. Check if date columns are stored as object
print("Order Date dtype:", records["Order Date"].dtype)
print("Ship Date dtype:", records["Ship Date"].dtype)

# 2. Check for leading/trailing spaces in Country
country_check = records["Country"].str.strip()
diff_spaces = (records["Country"] != country_check).sum()
print("Rows with trailing/leading spaces in Country:", diff_spaces)

# 3. Check for Unknown City entries
unknown_count = (records["shipping_city"] == "Unknown City").sum()
print("Rows with 'Unknown City':", unknown_count)

# 4. Check logic: Ship Date < Order Date
invalid_dates = (records["Ship Date"] < records["Order Date"]).sum()
print("Rows where Ship Date < Order Date:", invalid_dates)

# 5. Check for duplicates
duplicate_rows = records.duplicated().sum()
print("Duplicate rows:", duplicate_rows)

# 6. Check inconsistency in formating.

# Check unique Item Types (look for duplicates with different cases)
print("Unique Item Types:", records["Item Type"].unique())

# Lowercased version to spot duplicates
print("Unique Item Types (lowercased):", records["Item Type"].str.lower().unique())

# Check for country name inconsistencies
print("Unique Countries (sample 20):", records["Country"].unique()[:20])

# 7. Check for outliners in price.

# Check for zero or negative Units Sold
invalid_units = records[records["Units Sold"] <= 0]
print("Invalid Units Sold count:", len(invalid_units))

# Very large values (outliers) – for example > 10,000
outliers = records[records["Units Sold"] > 10000]
print("Outlier Units Sold count:", len(outliers))

# Check if totals match formulas
invalid_revenue = records[
    (records["Units Sold"] * records["Unit Price"]).round(2) != records["Total Revenue"].round(2)
]
invalid_cost = records[
    (records["Units Sold"] * records["Unit Cost"]).round(2) != records["Total Cost"].round(2)
]
invalid_profit = records[
    (records["Total Revenue"] - records["Total Cost"]).round(2) != records["Total Profit"].round(2)
]

print("Invalid Revenue rows:", len(invalid_revenue))
print("Invalid Cost rows:", len(invalid_cost))
print("Invalid Profit rows:", len(invalid_profit))

Order Date dtype: object
Ship Date dtype: object
Rows with trailing/leading spaces in Country: 22
Rows with 'Unknown City': 38
Rows where Ship Date < Order Date: 108
Duplicate rows: 0
Unique Item Types: ['Cosmetics' 'Vegetables' 'Baby Food' 'Cereal' 'Fruits' 'Clothes' 'Snacks'
 'Household' 'Office Supplies' 'Beverages' 'Personal Care' 'Meat']
Unique Item Types (lowercased): ['cosmetics' 'vegetables' 'baby food' 'cereal' 'fruits' 'clothes' 'snacks'
 'household' 'office supplies' 'beverages' 'personal care' 'meat']
Unique Countries (sample 20): ['Libya' 'Canada' 'Japan' 'Chad' 'Armenia' 'Eritrea' 'Montenegro'
 'Jamaica' 'Fiji' 'Togo' 'Greece' 'Sudan' 'Maldives' 'Estonia' 'Greenland'
 'Cape Verde' 'Senegal' 'Federated States of Micronesia' 'Bulgaria'
 'Algeria']
Invalid Units Sold count: 0
Outlier Units Sold count: 0
Invalid Revenue rows: 0
Invalid Cost rows: 0
Invalid Profit rows: 0


#### Conclusion

We validated the dataset across multiple dimensions to identify possible dirty data cases:

1. **Date columns**  
   - Both `Order Date` and `Ship Date` are already stored as `datetime64[ns]`.  
   - No rows found where `Ship Date < Order Date`.  

2. **Text field formatting**  
   - `Country` column contained 22 rows with trailing/leading whitespace, which can cause duplicates like `"Canada"` vs `"Canada "`.  
   - `Item Type` was consistent — unique values matched after lowercasing, so no case-duplication issues were found.  

3. **Missing/unmapped values**  
   - 38 rows had `"Unknown City"` in the derived `shipping_city` column. These represent missing mappings when converting from country to capital city.  

4. **Numeric validation**  
   - `Units Sold`: No invalid values (≤ 0) and no extreme outliers (> 10,000).  
   - Revenue, Cost, and Profit calculations matched expected formulas in all rows.  

5. **Duplicates**  
   - ✅ No duplicate rows detected.  

---

### ✅ Conclusion
The dataset is relatively clean and well-structured.  
The main grime cases are:  
- **22 rows** with extra spaces in `Country`,  
- **38 rows** with `"Unknown City"`, and  
- the need to ensure **dates are consistently treated as `datetime64[ns]`**.  

Other aspects (numeric consistency, duplicates, case sensitivity) did not show problems.


#### 7. Cleaning Rules
Lets apply the cleaning rules to the grime cases that was found above.
* Strip whitespace → Removed leading/trailing spaces in `Country` and other text columns (22 rows affected).  
* City mapping → Replaced `"Unknown City"` entries in `shipping_city` with a safe fallback (`"<Country> City"`)

In [35]:
def clean(df: pd.DataFrame) -> pd.DataFrame:
    out = df.copy()
    
    # 1) Strip whitespace in text/object columns
    for c in out.select_dtypes(include="object").columns:
        out[c] = out[c].str.strip()
    
    # 2) Replace "Unknown City" with "<Country> City" (only if column exists)
    if "shipping_city" in out.columns:
        mask = out["shipping_city"].eq("Unknown City")
        out.loc[mask, "shipping_city"] = out.loc[mask, "Country"] + " City"
    
    return out

# ---- Run before/after comparison
before_spaces = (records["Country"] != records["Country"].str.strip()).sum()
before_unknown = (records["shipping_city"] == "Unknown City").sum() if "shipping_city" in records.columns else "N/A"

cleaned = clean(records)

after_spaces = (cleaned["Country"] != cleaned["Country"].str.strip()).sum()
after_unknown = (cleaned["shipping_city"] == "Unknown City").sum() if "shipping_city" in cleaned.columns else "N/A"

print(f"Country whitespace → before: {before_spaces}, after: {after_spaces}")
print(f"Unknown City → before: {before_unknown}, after: {after_unknown}")
print("Row*Column after cleaning:", cleaned.shape)

Country whitespace → before: 22, after: 0
Unknown City → before: 38, after: 0
Row*Column after cleaning: (500, 15)


#### 8. Transformations

In [36]:
import pandas as pd

def transform(df: pd.DataFrame) -> pd.DataFrame:
    out = df.copy()

    # --- Dates: BEFORE ---
    print("\nBefore date dtype change:")
    print(out[['Order Date','Ship Date']].dtypes)
    print(out[['Order Date','Ship Date']].count())
    print(out['Order Date'].head(5))

    # Convert to datetime (coerce bad values -> NaT)
    out['Order Date'] = pd.to_datetime(out['Order Date'], errors='coerce')
    out['Ship Date']  = pd.to_datetime(out['Ship Date'],  errors='coerce')

    # --- Dates: AFTER ---
    print("\nAfter date dtype change:")
    print(out[['Order Date','Ship Date']].dtypes)
    print(out[['Order Date','Ship Date']].count())
    print(out['Order Date'].head(5))

    # --- Coupon transform (simple & deterministic) ---
    if 'coupon_code' not in out.columns:
        codes = ['NONE','SAVE05','SAVE10','SAVE15','SAVE20']
        out['coupon_code'] = out['Order ID'].astype(int).mod(len(codes)).map(dict(enumerate(codes)))

    # Parse % from coupon_code -> discount in [0,1]
    out['discount_pct'] = (
        out['coupon_code'].astype(str).str.upper().str.extract(r'(\d{1,2})')[0]
        .astype(float).fillna(0).clip(0,100).div(100)
    )

    # Derived fields (don’t overwrite originals)
    out['unit_price_after_discount'] = (out['Unit Price'] * (1 - out['discount_pct'])).round(2)
    out['discount_amount'] = (out['Unit Price'] - out['unit_price_after_discount']).round(2)
    out['net_revenue'] = (out['unit_price_after_discount'] * out['Units Sold']).round(2)
    out['net_profit']  = (out['net_revenue'] - out['Total Cost']).round(2)

    # Tiny summary
    print("\nCoupon summary (first 5):")
    print(out[['coupon_code','discount_pct','unit_price_after_discount','net_revenue','net_profit']].head(5))

    return out

# Run
df_transformed = transform(cleaned if 'cleaned' in locals() else records)


Before date dtype change:
Order Date    object
Ship Date     object
dtype: object
Order Date    500
Ship Date     500
dtype: int64
0    10/18/2014
1     11/7/2011
2    10/31/2016
3     4/10/2010
4     8/16/2011
Name: Order Date, dtype: object

After date dtype change:
Order Date    datetime64[ns]
Ship Date     datetime64[ns]
dtype: object
Order Date    500
Ship Date     500
dtype: int64
0   2014-10-18
1   2011-11-07
2   2016-10-31
3   2010-04-10
4   2011-08-16
Name: Order Date, dtype: datetime64[ns]

Coupon summary (first 5):
  coupon_code  discount_pct  unit_price_after_discount  net_revenue  \
0      SAVE05          0.05                     415.34   3507961.64   
1      SAVE10          0.10                     138.65    418445.70   
2      SAVE05          0.05                     242.52    367902.84   
3      SAVE20          0.20                     164.56    546668.32   
4        NONE          0.00                       9.33     91853.85   

   net_profit  
0  1283876.46  
1   1440

#### 9. Feature Engineering

In [37]:
import pandas as pd

def feature_engineering(df: pd.DataFrame) -> pd.DataFrame:
    out = df.copy()

    # 1) Add customer_id (unique row index starting at 1)
    out["customer_id"] = range(1, len(out) + 1)

    # 2) Days to ship (requires datetime columns)
    if pd.api.types.is_datetime64_any_dtype(out["Order Date"]) and pd.api.types.is_datetime64_any_dtype(out["Ship Date"]):
        out["days_to_ship"] = (out["Ship Date"] - out["Order Date"]).dt.days

    # 3) Profit margin percentage
    out["profit_margin_pct"] = ((out["Total Profit"] / out["Total Revenue"]) * 100).round(2)

    # Preview
    print(out[["customer_id", "Order Date", "Ship Date", "days_to_ship", "profit_margin_pct"]].head(5))
    return out

# Run
df_features = feature_engineering(df_transformed if "df_transformed" in locals() else records)

   customer_id Order Date  Ship Date  days_to_ship  profit_margin_pct
0            1 2014-10-18 2014-10-31            13              39.77
1            2 2011-11-07 2011-12-08            31              40.98
2            3 2016-10-31 2016-12-09            39              37.55
3            4 2010-04-10 2010-05-12            32              43.07
4            5 2011-08-16 2011-08-31            15              25.83


#### 10. 	Mini-Aggregation

In [38]:
# --- Step 10: Mini-Aggregation ---

# Aggregate total revenue by shipping city
revenue_by_city = (
    records.groupby("shipping_city")["Total Revenue"]
    .sum()
    .sort_values(ascending=False)
)

# Show top 3 cities
print("Top 3 cities by revenue:")
print(revenue_by_city.head(3))

# Optionally convert to dictionary (for JSON-friendly format)
revenue_dict = revenue_by_city.to_dict()

# Preview sample from dictionary
for k, v in list(revenue_dict.items())[:3]:
    print(f"{k}: {v:.2f}")

Top 3 cities by revenue:
shipping_city
Unknown City    36417096.86
Port Moresby    15629197.78
San José        15578320.59
Name: Total Revenue, dtype: float64
Unknown City: 36417096.86
Port Moresby: 15629197.78
San José: 15578320.59


#### 11. Serialization Checkpoint

In [39]:
# Save cleaned dataset to JSON
output_path = "../data/cleaned_records.json"
cleaned.to_json(output_path, orient="records", lines=True)

print(f"Cleaned data saved to {output_path}")

# Quick validation: read back the JSON
reloaded = pd.read_json(output_path, orient="records", lines=True)

print("Reloaded shape:", reloaded.shape)
print("First 2 rows from reloaded JSON:")
print(reloaded.head(2))

Cleaned data saved to ../data/cleaned_records.json
Reloaded shape: (500, 15)
First 2 rows from reloaded JSON:
                         Region Country   Item Type Sales Channel  \
0  Middle East and North Africa   Libya   Cosmetics       Offline   
1                 North America  Canada  Vegetables        Online   

  Order Priority  Order Date   Order ID   Ship Date  Units Sold  Unit Price  \
0              M  10/18/2014  686800706  10/31/2014        8446      437.20   
1              M   11/7/2011  185941302   12/8/2011        3018      154.06   

   Unit Cost  Total Revenue  Total Cost  Total Profit shipping_city  
0     263.33     3692591.20  2224085.18    1468506.02       Tripoli  
1      90.93      464953.08   274426.74     190526.34        Ottawa  


#### Step 12 — Soft Interview Reflection

Using functions throughout this lab was essential for writing clean and reusable code.  
Instead of repeating the same logic (e.g., converting dates, cleaning whitespace, or mapping cities), functions allowed me to encapsulate each task with clear inputs and outputs. This reduced errors, made debugging easier, and kept the notebook more organized. Functions also made testing more straightforward — I could run them on a sample dataset before applying to the full records. Most importantly, functions support collaboration: teammates can read, reuse, or extend my code without needing to understand every line in detail. Overall, functions increased clarity, reusability, and efficiency in the data-preprocessing workflow.

#### Data Dictionary (Summary)

- **Primary CSV fields**: Region, Country, Item Type, Sales Channel, Order Priority, Order Date, Order ID, Ship Date, Units Sold, Unit Price, Unit Cost, Total Revenue, Total Cost, Total Profit.  
  *Types:* object, int64, float64, datetime64[ns].  

- **Derived fields**:  
  * `shipping_city` → capital mapped from Country (with "Unknown City" handled).  
  * `customer_id` → synthetic unique ID (1…n).  
  * `days_since_purchase` → calculated days between Order Date and reference date.  
  * `coupon_code` + `discount_pct` → synthetically generated discounts.  
  * `unit_price_after_discount`, `discount_amount`, `net_revenue`, `net_profit` → computed from coupon rules.  

## Data Dictionary

| Field                     | Type          | Description                                                     | Source     |
|---------------------------|---------------|-----------------------------------------------------------------|------------|
| Region                    | object        | Geographic region of sales                                      | Primary CSV|
| Country                   | object        | Country of transaction (whitespace cleaned)                     | Primary CSV|
| Item Type                 | object        | Product category (e.g., Cosmetics, Vegetables, Baby Food)       | Primary CSV|
| Sales Channel             | object        | Sales channel (Online/Offline)                                  | Primary CSV|
| Order Priority            | object        | Priority code (H = High, M = Medium, L = Low, C = Critical)     | Primary CSV|
| Order Date                | datetime64[ns]| Date when the order was placed                                  | Primary CSV|
| Order ID                  | int64         | Unique identifier for the order                                 | Primary CSV|
| Ship Date                 | datetime64[ns]| Date when the order was shipped                                 | Primary CSV|
| Units Sold                | int64         | Quantity sold                                                   | Primary CSV|
| Unit Price                | float64       | Price per unit                                                  | Primary CSV|
| Unit Cost                 | float64       | Cost per unit                                                   | Primary CSV|
| Total Revenue             | float64       | Computed: Units Sold × Unit Price                               | Primary CSV|
| Total Cost                | float64       | Computed: Units Sold × Unit Cost                                | Primary CSV|
| Total Profit              | float64       | Computed: Total Revenue − Total Cost                            | Primary CSV|
| shipping_city             | object        | Capital city mapped from Country (Unknown City filled/fixed)    | Derived    |
| customer_id               | int64         | Synthetic unique customer ID (1…n)                              | Synthetic  |
| days_since_purchase       | int64         | Days between Order Date and reference date (2025-09-30)         | Derived    |
| coupon_code               | object        | Synthetic coupon generated from Order ID                        | Synthetic  |
| discount_pct              | float64       | Numeric discount parsed from coupon_code (0–0.20)               | Derived    |
| unit_price_after_discount | float64       | Unit price after applying discount                              | Derived    |
| discount_amount           | float64       | Amount reduced per unit due to coupon                           | Derived    |
| net_revenue               | float64       | Revenue after discounts (unit_price_after_discount × Units Sold)| Derived    |
| net_profit                | float64       | Profit after discounts (net_revenue − Total Cost)               | Derived    |
